### PR TITLE
Response map HTTP filter for custom upstream responses

### LINKED ISSUE #12883: Custom upstrem/backend response
I want to rewrite the 4xx/5xx returned by the upstrem/backend service.
I see that envoy 1.15.0 has been able to rewrite the error of envoy itself through local_reply_config, but it seems that it cannot rewrite the error returned by upstrem/backend. The function of local_reply_config is very good and can replace haproxy as the front end lb. It would be even better if could rewrite upstrem/backend.

I don’t know if my configuration is wrong, my configuration：

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: local-reply
  namespace: istio-system
spec:
  configPatches:
  - applyTo: NETWORK_FILTER # http connection manager is a filter in Envoy
    match:
      context: GATEWAY
      listener:
        filterChain:
          filter:
            name: "envoy.http_connection_manager"
    patch:
      operation: MERGE
      value:
        typed_config:
          "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
          local_reply_config:
            mappers:
            - filter:
                status_code_filter:
                  comparison:
                    op: EQ
                    value:
                      default_value: 503
                      runtime_key: key_b
              status_code: 200
              body:
                inline_string: "no health server"
              body_format_override:
                json_format:
                  status: "503"
                  message: "%LOCAL_REPLY_BODY%"
                  request_path: "%REQ(:path)%"
```

```
[root@k8s-20 ~]# curl -i http://test.domain.com/503/test/path                                                                                            
HTTP/1.1 503 Service Unavailable
server: nginx/1.15.6
date: Sat, 29 Aug 2020 03:51:36 GMT
content-type: text/html
content-length: 203
x-request-id: 3c4a04cd-c4d3-4819-af76-256b6c905a0b

<html>
<head><title>503 Service Temporarily Unavailable</title></head>
<body>
<center><h1>503 Service Temporarily Unavailable</h1></center>
<hr><center>openresty/1.15.8.2</center>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by esmet**
Support custom upstream responses using a new `response_map` filter.

The response map filter behaves like the local reply code, but on the HTTP filter chain instead. This allows for modification of responses generated both locally and upstream.

The motivation and implementation strategy for this change is captured in https://docs.google.com/document/d/1THyRnepx89FbK2VVaqoEHXv2knbm_lsLx2XgDZ7FxDw/edit?usp=sharing

Commit Message: 
Additional Description:
Risk Level: medium (new filter, refactored common code)
Testing: unit tests done, integration tests still needed.
Docs Changes: NEEDED
Release Notes: NEEDED
Platform Specific Features:
Fixes #12883

### CODE REVIEW

File: source/common/response_map/response_map.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+   * @param stream_info supplies the information about streams required by filters.
+   * @param code status code.
+   * @param body response body.
+   * @param content_type response content_type.
+   */
+  virtual void rewrite(const Http::RequestHeaderMap* request_headers,
+                       Http::ResponseHeaderMap& response_headers,
+                       StreamInfo::StreamInfo& stream_info, Http::Code& code, std::string& body,
+                       absl::string_view& content_type) const PURE;
+
```
Comment (esmet) [inline source/common/response_map/response_map.h:28]: Still needs comments

File: source/common/response_map/response_map.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  virtual bool match(const Http::RequestHeaderMap* request_headers,
+                     const Http::ResponseHeaderMap& response_headers,
+                     const Http::ResponseTrailerMap* response_trailers,
+                     StreamInfo::StreamInfo& stream_info) const PURE;
+};
+
+using ResponseMapPtr = std::unique_ptr<ResponseMap>;
+
+/**
+ * Access log filter factory that reads from proto.
```
Comment (esmet) [inline source/common/response_map/response_map.h:38]: This comment could be improved

File: source/extensions/filters/http/response_map/BUILD [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    "//bazel:envoy_build_system.bzl",
+    "envoy_cc_extension",
+    "envoy_cc_library",
+    "envoy_extension_package",
+)
+
+licenses(["notice"])  # Apache 2
+
+# L7 HTTP filter that implements response modifications using matchers.
+# Public docs: docs/root/configuration/http_filters/response_map_filter.rst
```
Comment (esmet) [inline source/extensions/filters/http/response_map/BUILD:11]: This doesn't exist yet.

File: source/extensions/filters/http/response_map/response_map_filter.cc [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  return per_route_config;
+}
+
+/*
+ * Called if there are any downstream headers to decode from the client or a previous filter.
+ *
+ * Not guaranteed to be called. In particular, if Envoy sends a local reply with HTTP 426
+ * as an upgrade response for HTTP/1.0 requests, we may never decode any of the client's
+ * original headers, because the request was rejected at the initial HTTP/1.0 line.
```
Comment (esmet) [inline source/extensions/filters/http/response_map/response_map_filter.cc:61]: This comment might be better off above the original `decodeHeaders` function declaration.

File: source/extensions/filters/http/response_map/response_map_filter.cc [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  // or to a non-null per-route config. Guard against it anyway.
+  if (response_map_ == nullptr) {
+    ENVOY_LOG(trace,
+              "response map filter: doRewrite has no response_map_ to rewrite with, doing nothing");
+    return;
+  }
+
+  const ResponseMap::ResponseMapPtr& response_map = *response_map_;
+
+  // Fill in new_body and new_content_type using the response map rewrite.
```
Comment (esmet) [inline source/extensions/filters/http/response_map/response_map_filter.cc:203]: Extend this comment to mention that we capture the rewritten `code` here per the method signature, but we don't actually do anything with the new value.

File: source/extensions/filters/http/response_map/response_map_filter.cc [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+  std::string new_body;
+  absl::string_view new_content_type;
+  auto code = static_cast<Http::Code>(Http::Utility::getResponseStatus(*response_headers_));
+  response_map->rewrite(request_headers_, *response_headers_, encoder_callbacks_->streamInfo(),
+                        code, new_body, new_content_type);
+
+  // Encoding buffer may be null here even if we saw data in encodeData above. This
+  // happens when sendLocalReply sends a response downstream. By adding encoded data
+  // here, we do successfully override the sendLocalReply body, but it's not clear
+  // how or why.
```
Comment (esmet) [inline source/extensions/filters/http/response_map/response_map_filter.cc:215]: We should get some clarity around this condition here and rewrite the comment to explain what's going on instead of asking questions.

File: source/extensions/filters/http/response_map/response_map_filter.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override {
+    return Http::FilterDataStatus::Continue;
+  }
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap&) override {
+    return Http::FilterTrailersStatus::Continue;
+  }
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override {
+    decoder_callbacks_ = &callbacks;
```
Comment (esmet) [inline source/extensions/filters/http/response_map/response_map_filter.h:98]: We never use this, so it should probably be removed.

File: test/common/stream_info/test_util.h [hunk 8]
```diff
@@ -38,6 +38,7 @@ class TestStreamInfo : public StreamInfo::StreamInfo {
   const absl::optional<std::string>& responseCodeDetails() const override {
     return response_code_details_;
   }
+  void setResponseCode(uint32_t code) override { response_code_ = code; }
```
Comment (esmet) [inline test/common/stream_info/test_util.h:41]: I would be in favor of moving this small addition to another pull request if others think that's better.

File: api/envoy/config/response_map/v3/response_map.proto [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: ResponseMap]
+// Response map filter :ref:`configuration overview <config_response_map>`.
+
+// The configuration to filter and change HTTP responses.
+// [#next-free-field: 6]
+message ResponseMapper {
+  // Filter to determine if this mapper should apply.
+  accesslog.v3.AccessLogFilter filter = 1 [(validate.rules).message = {required: true}];
```
Comment (alyssawilk) [inline api/envoy/config/response_map/v3/response_map.proto:28]: @lizan can I ask you to do an API pass before I dig into the code?
I think a lot of the components of access logs apply but not all do - not sure how much API refactor we want to do here and how much we can just do via standard metadata matcher.

File: api/envoy/config/core/v3/substitution_format_string.proto [hunk 10]
```diff
@@ -61,6 +64,12 @@ message SubstitutionFormatString {
     //  }
     //
     google.protobuf.Struct json_format = 2 [(validate.rules).message = {required: true}];
+
+    // A DataSource variant for text_format.
+    DataSource text_format_source = 5;
+
+    // The empty format string.
+    bool empty_format = 6 [(validate.rules).bool = {const: true}];
```
Comment (lizan) [inline api/envoy/config/core/v3/substitution_format_string.proto:72]: I don't see where this is used, shall we just relax the min_len: 1 constraint on text_format if we want an empty format string?

Comment (esmet) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: We use it here https://github.com/envoyproxy/envoy/pull/14221/files#diff-7a3f06d418ccdf0de0f51a473b6d3450baa5faf9ac8362aa6fe33bc1284b5152R27

We could just relax the constraint, but my initial idea was to avoid making changes to an existing field and figured the explicit "empty" option was clear enough.

Comment (lizan) [reply api/envoy/config/core/v3/substitution_format_string.proto:72]: I would slightly prefer relax the existing one, or we can have a field `fixed_text` which just do what PlainStringFormatterImpl does.

File: api/envoy/config/response_map/v3/response_map.proto (hunk 9 above)
Comment (lizan) [reply api/envoy/config/response_map/v3/response_map.proto:28]: @snowp can you weigh in here re: the new matcher API fits here? I think the access log filter works well for this purpose but just want to double check.

**Review by lizan — COMMENTED** [review]
can we split the matcher / substitution part to another PR?

**Comment by esmet** [comment]
> can we split the matcher / substitution part to another PR?

Yes, will do. I will also look into replacing `empty_format` with `fixed_text` and passing that text to PlainStringFormatterImpl.

File: api/envoy/config/response_map/v3/response_map.proto [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+option java_outer_classname = "ResponseMapProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: ResponseMap]
+// Response map filter :ref:`configuration overview <config_response_map>`.
+
+// The configuration to filter and change HTTP responses.
+// [#next-free-field: 6]
+message ResponseMapper {
```
Comment (lizan) [inline api/envoy/config/response_map/v3/response_map.proto:26]: This isn't different from [ResponseMapper](https://github.com/envoyproxy/envoy/blob/master/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L620) in HCM, why not just use it?

File: api/envoy/config/response_map/v3/response_map.proto [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  core.v3.SubstitutionFormatString body_format_override = 4;
+
+  // HTTP headers to add to a matched response. This allows the response mapper to append, to add
+  // or to override headers of any matched response before it is sent to a downstream client.
+  repeated core.v3.HeaderValueOption headers_to_add = 5
+      [(validate.rules).repeated = {max_items: 1000}];
+}
+
+// The configuration to customize HTTP responses read by Envoy.
+message ResponseMap {
```
Comment (lizan) [inline api/envoy/config/response_map/v3/response_map.proto:47]: This is same as LocalReplyConfig in HCM as well?

**Review by lizan — CHANGES_REQUESTED** [review]

File: api/envoy/config/response_map/v3/response_map.proto [hunk 13]
```diff
@@ -0,0 +1,89 @@
+syntax = "proto3";
```
[the remark this answers is not in the document]
Comment (esmet) [reply api/envoy/config/response_map/v3/response_map.proto:1]: The intention here is for this to be non-extension config (for source/common/response_map/*) and to have separate extension config for the response map filter (source/extensions/http/filters/response_map/...).

Is this idea correct or misguided?

File: api/envoy/config/response_map/v3/response_map.proto (hunk 12 above)
Comment (esmet) [reply api/envoy/config/response_map/v3/response_map.proto:47]: Yep. From the comment above, o intend for there to be a common proto shared across use cases and that we should (eventually) migrate HCM to use this common type. I would appreciate thoughts/guidance on a migration strategy.

File: api/envoy/config/response_map/v3/response_map.proto (hunk 9 above)
Comment (snowp) [reply api/envoy/config/response_map/v3/response_map.proto:28]: I would think that this could be accomplished with the new matching API, though we would most likely spend some time to reach parity in terms of supporting all the same conditionals that the access log filter supports. The matching API would offer some benefit in terms of better performance when lots of matchers are used, due to being able to structure it more as a tree than as a list of complex matchers. 

I think this falls into the category that we'd use the matching API if it was available today but since that work is ongoing it becomes a question of time lines. Maybe @mattklein123 has thoughts on whether we're okay with relying on the access log filters here.

I could imagine the access log filter eventually being subsumed by the new matching work, but that would be a longer term thing

File: api/envoy/config/response_map/v3/response_map.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (esmet) [reply api/envoy/config/response_map/v3/response_map.proto:1]: I looked harder at how the API is organized and I think I understand. My feeling now is that this should live under `api/envoy/extensions/common/reponse_map` since I intend for it to be shared, eventually, by both the HCM and ResponseMap filter. Does that make sense?

File: api/envoy/config/response_map/v3/response_map.proto (hunk 11 above)
Comment (esmet) [reply api/envoy/config/response_map/v3/response_map.proto:26]: I captured my thoughts here https://docs.google.com/document/d/1THyRnepx89FbK2VVaqoEHXv2knbm_lsLx2XgDZ7FxDw and added a link to the PR description

Comment (lizan) [reply api/envoy/config/response_map/v3/response_map.proto:26]: Thanks, can you give us comment access to the doc?

**Review by lizan — CHANGES_REQUESTED** [review]
Thanks for putting in the doc, let's discuss about design options there first and come back to the actual implementation.

The major concern I have is this configuration model as a filter could be very confusing, as the local reply by routers and filters are also going through encoder filter so it will affected by both LocalReplyConfig and this filter. Having these configuration in separate places doesn't seems a good design.  cc @alyssawilk @envoyproxy/api-shepherds

File: api/envoy/config/response_map/v3/response_map.proto (hunk 11 above)
Comment (esmet) [reply api/envoy/config/response_map/v3/response_map.proto:26]: Done

**Comment by htuch** [comment]
Could we just add an option to the HCM response mapper to also apply to upstream responses?

**Comment by lizan** [comment]
> Could we just add an option to the HCM response mapper to also apply to upstream responses?

Yeah I'm leaning to this as well.

**Comment by mattklein123** [comment]
> Yeah I'm leaning to this as well.

+1. I was just about to start looking through the doc, but I'm strongly against this amount of duplication and it seems like we can just extend the mapper to work with upstream responses also. Can we potentially re-orient the doc to that solution and then talk through any issues that come up in the doc?

**Comment by esmet** [comment]
> > Yeah I'm leaning to this as well.
> 
> +1. I was just about to start looking through the doc, but I'm strongly against this amount of duplication and it seems like we can just extend the mapper to work with upstream responses also. Can we potentially re-orient the doc to that solution and then talk through any issues that come up in the doc?

I'll look into how we could do this in the HCM and update the doc with the new research. When I originally designed this code, I was thinking that the filter chain would be invoked for all sendLocalReply calls but @lizan showed me that this isn't true in all cases (in particular, stream timeouts etc).

**Comment by esmet** [comment]
I updated the doc. It now tries to focus much more on the problem statement and possible solutions and less on the particulars of this pull request or even code in general. I will now start doing more code research and will update the doc with those findings as they become relevant.

**Comment by alyssawilk** [comment]
I think that having the HCM matcher be able to alter upstream responses is worth doing, especially for the common case where some server has global response tweaks they want to apply which can't be done via existing config knobs, but there's still the problem of per-route config to deal with.  That feels like a better fit for the RouteEntryImplBase::perFilterConfig which currently requires an explicit filter, though if the @envoyproxy/api-shepherds have a way to generalize that it could address that particular concern.

File: api/envoy/config/response_map/v3/response_map.proto (hunk 9 above)
Comment (alyssawilk) [reply api/envoy/config/response_map/v3/response_map.proto:28]: Yeah, I'd initially hoped the new matcher code could be used, since using access filters just strikes me as suboptimal and I'd like the perf wins.  Looks like it'd also do away with much of the shared code, so probably the way to go (if we opt to have a standalone filter for per-route) is stick with this for now, and move both the HCM and filter matchers over to the new API when it's more feature complete.

**Comment by mattklein123** [comment]
> That feels like a better fit for the RouteEntryImplBase::perFilterConfig which currently requires an explicit filter, though if the @envoyproxy/api-shepherds have a way to generalize that it could address that particular concern.

Yeah agreed that we should have per-route config in addition to generic matching, though I think this is also an issue for local replies, since in many cases a local reply may be generated when a route is available. I would need to look into it in more detail but I wonder if we could just have a non-filter key for override config that could be looked up?

**Comment by alyssawilk** [comment]
If you're OK with a non-filter key that works for me.  Alternately @esmet suggested offline that (with some possible tweaking to the per-filter-config code) we could arguably use the HCM network filter name (it is a filter, just not L7 =P) instead of an arbitrary string.  WDYT?

**Comment by mattklein123** [comment]
> If you're OK with a non-filter key that works for me. Alternately @esmet suggested offline that (with some possible tweaking to the per-filter-config code) we could arguably use the HCM network filter name (it is a filter, just not L7 =P) instead of an arbitrary string. WDYT?

Either SGTM. I agree the latter sounds interesting and we could have a generic config message that might eventually allow overriding various pieces of HCM behavior on a per-route basis.

**Comment by lizan** [comment]
> I think that having the HCM matcher be able to alter upstream responses is worth doing, especially for the common case where some server has global response tweaks they want to apply which can't be done via existing config knobs, but there's still the problem of per-route config to deal with. That feels like a better fit for the RouteEntryImplBase::perFilterConfig which currently requires an explicit filter, though if the @envoyproxy/api-shepherds have a way to generalize that it could address that particular concern.

If we're not implementing the response mapper in a filter, then it can be a field in Route message and shouldn't be part of [`typed_per_filter_config`](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/route/v3/route_components.proto#L252) anyway? Then it won't use `RouteEntryImplBase::perFilterConfig` either.

**Comment by alyssawilk** [comment]
if we're not doing it in a filter, don't we want it all in the HCM/FM for behavioral consistency? it's not clear to me if we'd need the route specific logic if Envoy sent a local reply late in the response process, but the FM is the only entity which sees the direct responses.

**Comment by esmet** [comment]
This is still on my plate. The next thing that I plan to do is to try moving the response mapping code out of an HTTP extension and into the HCM itself.

**Comment by esmet** [comment]
I submitted a new draft PR that implements this behavior in the FilterManager https://github.com/envoyproxy/envoy/pull/14926 feedback welcome.

I'm keeping branch/PR unchanged so we have the HTTP filter implementation to compare to, and preserve history.

**Comment by alexandru-ersenie** [comment]
@esmet : The way i look at it is that your work is still in progress, and there is no way to control the error pages for example, returned by the upstream. 

As we are using envoy as a proxy to the Istio Mesh, my understanding is we'd have to wait a little bitt longer, am i correct to assume that ?

Thanks !

**Comment by esmet** [comment]
@alexandru-ersenie yes, this work is still in progress and I need to get it to a place where we achieve consensus around how it will be implemented. I will hopefully be able to pick this up again soon.
